Use a less restrictive iterator definition in hpx::lcos::detail::future_iterator_traits

Currently, HPX defines future iterators as something which has an iterator_category member type. See hpx/lcos/future.hpp:293...

    template <typename Iter, typename Enable = void>
    struct future_iterator_traits
    {};

    template <typename Iterator>
    struct future_iterator_traits<Iterator,
        typename hpx::util::always_void<
            typename Iterator::iterator_category
        >::type>
    {
        typedef
            typename std::iterator_traits<Iterator>::value_type
            type;

        typedef hpx::traits::future_traits<type> traits_type;
    };

This definition is too restrictive as it discards the single most common type of C++ iterator: pointers. Which means, for example, that wait_all cannot operate on an std::array of futures, ie `hpx::wait_all(array.cbegin(), array.cend())` will turn into a long trail of compiler vomit.

I think that this should be corrected by using std::iterator_traits all the way up instead. Here is a proposal:

    template <typename Iterator>
    struct future_iterator_traits<Iterator,
        typename hpx::util::always_void<
            typename std::iterator_traits<Iterator>::value_type
        >::type>
    {
        typedef
            typename std::iterator_traits<Iterator>::value_type
            type;

        typedef hpx::traits::future_traits<type> traits_type;
    };

My pleasure!

This code should probably compile, but doesn't (on GCC 4.8, Linux x86_64):

```
#include <array>

#include <hpx/hpx_main.hpp>
#include <hpx/include/future.hpp>
#include <hpx/lcos/wait_all.hpp>

int main() {
   std::array<hpx::shared_future<int>, 1> future_array { hpx::make_ready_future(0) };
   hpx::wait_all(future_array.cbegin(), future_array.cend());
}
```

Relevant compiler output:

```
In file included from /usr/local/include/hpx/traits/acquire_shared_state.hpp:11:0,
                 from /usr/local/include/hpx/lcos/future.hpp:13,
                 from /usr/local/include/hpx/include/future.hpp:9,
                 from /home/hadrien/Bureau/Programmation/HPX/Tests/Futurization/future_iterator_crasher.cpp:3:
/usr/local/include/hpx/traits/future_access.hpp: In instantiation of ‘struct hpx::traits::detail::shared_state_ptr_for<const hpx::lcos::shared_future<int>*>’:
/usr/local/include/hpx/lcos/wait_all.hpp:396:15:   required from ‘void hpx::lcos::wait_all(Ts&& ...) [with Ts = const hpx::lcos::shared_future<int>*, const hpx::lcos::shared_future<int>*]’
/home/hadrien/Bureau/Programmation/HPX/Tests/Futurization/future_iterator_crasher.cpp:8:60:   required from here
/usr/local/include/hpx/traits/future_access.hpp:42:16: error: no type named ‘type’ in ‘struct hpx::traits::future_traits<const hpx::lcos::shared_future<int>*, void>’
         struct shared_state_ptr_for
                ^
In file included from /home/hadrien/Bureau/Programmation/HPX/Tests/Futurization/future_iterator_crasher.cpp:4:0:
/usr/local/include/hpx/lcos/wait_all.hpp: In instantiation of ‘void hpx::lcos::wait_all(Ts&& ...) [with Ts = const hpx::lcos::shared_future<int>*, const hpx::lcos::shared_future<int>*]’:
/home/hadrien/Bureau/Programmation/HPX/Tests/Futurization/future_iterator_crasher.cpp:8:60:   required from here
/usr/local/include/hpx/lcos/wait_all.hpp:396:15: error: no type named ‘type’ in ‘struct hpx::traits::detail::shared_state_ptr_for<const hpx::lcos::shared_future<int>*>’
             > result_type;
               ^
/usr/local/include/hpx/lcos/wait_all.hpp:397:53: error: no type named ‘type’ in ‘struct hpx::traits::detail::shared_state_ptr_for<const hpx::lcos::shared_future<int>*>’
         typedef detail::wait_all_frame<result_type> frame_type;
                                                     ^
```

Trying this branch as-is, the test failed with this GCC warning, which was promoted into an error by Werror:

```
/home/hadrien/Bureau/Programmation/HPX/hpx/tests/regressions/lcos/wait_all_std_array_2035.cpp: In function ‘int hpx_startup::user_main()’:
/home/hadrien/Bureau/Programmation/HPX/hpx/tests/regressions/lcos/wait_all_std_array_2035.cpp:19:89: error: missing braces around initializer for ‘std::__array_traits<hpx::lcos::shared_future<int>, 1ul>::_Type {aka hpx::lcos::shared_future<int> [1]}’ [-Werror=missing-braces]
    std::array<hpx::shared_future<int>, 1> future_array { hpx::make_ready_future(0) };
                                                                                         ^
```

Investigation suggests that there might be a [bug in this GCC warning](http://stackoverflow.com/questions/11551986/gcc-missing-braces-around-initializer/). Anyhow, after applying this patch...

```diff
--- a/tests/regressions/lcos/wait_all_std_array_2035.cpp
+++ b/tests/regressions/lcos/wait_all_std_array_2035.cpp
@@ -16,7 +16,7 @@
 
 int main()
 {
-   std::array<hpx::shared_future<int>, 1> future_array { hpx::make_ready_future(0) };
+   std::array<hpx::shared_future<int>, 1> future_array {{ hpx::make_ready_future(0) }};
    hpx::wait_all(future_array.cbegin(), future_array.cend());
 
    return 0;
```

...the test now passes.

Adding extra braces alters a

--- a/hpx/lcos/future.hpp
+++ b/hpx/lcos/future.hpp
@@ -297,7 +297,7 @@ namespace hpx { namespace lcos { namespace detail
     template <typename Iterator>
     struct future_iterator_traits<Iterator,
         typename hpx::util::always_void<
-            typename Iterator::iterator_category
+            typename std::iterator_traits<Iterator>::value_type
         >::type>
     {
         typedef

--- a/tests/regressions/lcos/CMakeLists.txt
+++ b/tests/regressions/lcos/CMakeLists.txt
@@ -127,7 +127,7 @@ foreach(compile_test ${compile_tests})
     FOLDER "Tests/Regressions/LCOs/CompileOnly")
 
     # add a custom target for this example
-    add_hpx_pseudo_target(tests.unit.actions.${compile_test})
+    add_hpx_pseudo_target(tests.regressions.lcos_dir.${compile_test})
 
     # make pseudo-targets depend on master pseudo-target
     add_hpx_pseudo_dependencies(tests.regressions.lcos_dir

--- a/tests/regressions/lcos/CMakeLists.txt
+++ b/tests/regressions/lcos/CMakeLists.txt
@@ -107,3 +107,34 @@ add_hpx_test(
   EXECUTABLE after_588
   LOCALITIES 2
   THREADS_PER_LOCALITY 4)
+
+# compile only tests
+set(compile_tests
+    wait_all_std_array_2035
+   )
+
+foreach(compile_test ${compile_tests})
+  set(sources
+      ${compile_test}.cpp)
+
+  source_group("Source Files" FILES ${sources})
+
+  add_hpx_unit_compile_test(
+    "lcos_regressions"
+    ${compile_test}
+    SOURCES ${sources}
+    ${${compile_test}_FLAGS}
+    FOLDER "Tests/Regressions/LCOs/CompileOnly")
+
+    # add a custom target for this example
+    add_hpx_pseudo_target(tests.unit.actions.${compile_test})
+
+    # make pseudo-targets depend on master pseudo-target
+    add_hpx_pseudo_dependencies(tests.regressions.lcos_dir
+      tests.regressions.lcos_dir.${compile_test})
+
+    # add dependencies to pseudo-target
+    add_hpx_pseudo_dependencies(tests.regressions.lcos_dir.${compile_test}
+      "tests.regressions.lcos_dir.${compile_test}")
+
+endforeach()

--- a/tests/regressions/lcos/wait_all_std_array_2035.cpp
+++ b/tests/regressions/lcos/wait_all_std_array_2035.cpp
@@ -1,0 +1,24 @@
+//  Copyright (c) 2016 Hadrian G. (a.k.a. Neolander)
+//
+//  Distributed under the Boost Software License, Version 1.0. (See accompanying
+//  file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+// This compile-only test case verifies that #2035 remains fixed
+
+#include <hpx/config.hpp>
+
+#if defined(HPX_HAVE_CXX11_STD_ARRAY)
+#include <hpx/hpx_main.hpp>
+#include <hpx/include/future.hpp>
+#include <hpx/lcos/wait_all.hpp>
+
+#include <array>
+
+int main()
+{
+   std::array<hpx::shared_future<int>, 1> future_array { hpx::make_ready_future(0) };
+   hpx::wait_all(future_array.cbegin(), future_array.cend());
+
+   return 0;
+}
+#endif

--- a/tests/regressions/lcos/wait_all_std_array_2035.cpp
+++ b/tests/regressions/lcos/wait_all_std_array_2035.cpp
@@ -16,9 +16,11 @@
 
 int main()
 {
-   std::array<hpx::shared_future<int>, 1> future_array { hpx::make_ready_future(0) };
-   hpx::wait_all(future_array.cbegin(), future_array.cend());
+    std::array<hpx::shared_future<int>, 1> future_array
+        {{ hpx::make_ready_future(0) }};
 
-   return 0;
+    hpx::wait_all(future_array.cbegin(), future_array.cend());
+
+    return 0;
 }
 #endif

